dandi/delete.py

```diff
@@ -0,0 +1,220 @@
+from dataclasses import dataclass, field
+from operator import attrgetter
+from pathlib import Path, PurePosixPath
+from typing import Iterable, Iterator, NamedTuple, Optional, Set, Tuple
+
+import click
+import requests
+
+from .consts import known_instances
+from .dandiapi import DandiAPIClient
+from .dandiarchive import parse_dandi_url
+from .exceptions import NotFoundError
+from .utils import is_url
+
+
+class RemoteAsset(NamedTuple):
+    dandiset_id: str
+    version_id: str
+    asset_id: str
+    path: str
+
+
+@dataclass
+class Deleter:
+    client: Optional[DandiAPIClient] = None
+    dandiset_id: Optional[str] = None
+    deleting_dandiset: bool = False
+    remote_assets: Set[RemoteAsset] = field(default_factory=set)
+
+    def __bool__(self) -> bool:
+        return self.deleting_dandiset or bool(self.remote_assets)
+
+    def set_dandiset(self, api_url: str, dandiset_id: str) -> None:
+        if self.client is None:
+            self.client = DandiAPIClient(api_url)
+            self.client.dandi_authenticate()
+            self.dandiset_id = dandiset_id
+            try:
+                self.client.get_dandiset(dandiset_id, "draft")
+            except requests.HTTPError as e:
+                if e.response.status_code == 404:
+                    raise NotFoundError(f"Dandiset {dandiset_id} not found on server")
+                else:
+                    raise
+        elif not is_same_url(self.client.api_url, api_url):
+            raise ValueError("Cannot delete assets from multiple API instances at once")
+        elif self.dandiset_id != dandiset_id:
+            raise ValueError("Cannot delete assets from multiple Dandisets at once")
+
+    def register_dandiset(self, api_url: str, dandiset_id: str) -> None:
+        self.set_dandiset(api_url, dandiset_id)
+        self.deleting_dandiset = True
+
+    def register_asset(
+        self, api_url: str, dandiset_id: str, version_id: str, asset_path: str
+    ) -> None:
+        if version_id != "draft":
+            raise ValueError("Cannot delete assets from published versions")
+        self.set_dandiset(api_url, dandiset_id)
+        asset = self.client.get_asset_bypath(dandiset_id, version_id, asset_path)
+        if asset is None:
+            raise NotFoundError(
+                f"Asset at path {asset_path!r} not found in Dandiset {dandiset_id}"
+            )
+        self.remote_assets.add(
+            RemoteAsset(dandiset_id, version_id, asset["asset_id"], asset["path"])
+        )
+
+    def register_asset_folder(
+        self, api_url: str, dandiset_id: str, version_id: str, folder_path: str
+    ) -> None:
+        if version_id != "draft":
+            raise ValueError("Cannot delete assets from published versions")
```

Comment:
should we just rely on dandi-api to disallow such action and client to not care?  sorry if I forgot if we ran into some stumbling point on that or what -- if to keep, a comment should be added on why we need this client side logic